Terminal.Gui/View/Layout/Dim.cs

```diff
@@ -0,0 +1,766 @@
+
+    /// <summary>
+    ///     Gets the maximum dimension the View's ContentSize will be fit to. NOT CURRENTLY SUPPORTED.
+    /// </summary>
+    public Dim MaximumContentDim { get; } = maximumContentDim;
+
+    /// <summary>
+    ///     Gets the minimum dimension the View's ContentSize will be constrained to.
+    /// </summary>
+    public Dim MinimumContentDim { get; } = minimumContentDim;
+
+    /// <summary>
+    ///     Gets the style of the DimAuto.
+    /// </summary>
+    public DimAutoStyle Style { get; } = style;
+
+    /// <inheritdoc/>
+    public override string ToString () { return $"Auto({Style},{MinimumContentDim},{MaximumContentDim})"; }
+
+    internal override int Calculate (int location, int superviewContentSize, View us, Dimension dimension)
+    {
+        if (us == null)
+        {
+            return MaximumContentDim?.Anchor (0) ?? 0;
+        }
+
+        var textSize = 0;
+        var subviewsSize = 0;
+
+        int autoMin = MinimumContentDim?.Anchor (superviewContentSize) ?? 0;
+
+        if (superviewContentSize < autoMin)
+        {
+            Debug.WriteLine ($"WARNING: DimAuto specifies a min size ({autoMin}), but the SuperView's bounds are smaller ({superviewContentSize}).");
+
+            //return superviewContentSize;
+        }
+
+        if (Style.HasFlag (DimAutoStyle.Text))
+        {
+            textSize = int.Max (autoMin, dimension == Dimension.Width ? us.TextFormatter.Size.Width : us.TextFormatter.Size.Height);
+        }
+
+        if (Style.HasFlag (DimAutoStyle.Content))
+        {
+            if (us._contentSize is { })
+            {
+                subviewsSize = dimension == Dimension.Width ? us.ContentSize.Width : us.ContentSize.Height;
+            }
+            else
+            {
+                // TODO: AnchorEnd needs work
+                // TODO: If _min > 0 we can SetRelativeLayout for the subviews?
+                subviewsSize = 0;
+
+                if (us.Subviews.Count > 0)
+                {
+                    for (var i = 0; i < us.Subviews.Count; i++)
+                    {
+                        View v = us.Subviews [i];
+                        bool isNotPosAnchorEnd = dimension == Dimension.Width ? v.X is not PosAnchorEnd : v.Y is not PosAnchorEnd;
+
+                        //if (!isNotPosAnchorEnd)
+                        //{
+                        //    v.SetRelativeLayout(dimension == Dimension.Width ? (new Size (autoMin, 0)) : new Size (0, autoMin));
+                        //}
+
+                        if (isNotPosAnchorEnd)
+                        {
+                            int size = dimension == Dimension.Width ? v.Frame.X + v.Frame.Width : v.Frame.Y + v.Frame.Height;
+
+                            if (size > subviewsSize)
+                            {
+                                subviewsSize = size;
+                            }
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
Holy cyclomatic complexity, Batman!

This stuff should get refactored to reduce nesting.

ReSharper should be able to help out with that a lot.

Usually easiest to work from the outside in, if trying to do it on the outermost first and letting it turn it into a switch doesn't have a nice result.

But this whole nested control flow can be one big single switch with pattern matching, performed in the reverse order (most specific first because order matters), in half or fewer lines, a single level of nesting, and without having to follow a binary tree for a human to parse it.

Reply:
See new comment:

```cs
// TODO: This whole body of code is a WIP (for https://github.com/gui-cs/Terminal.Gui/pull/3451).
```

Reply:
I think I put a WIP tag in for resharper, didn't I? If not, I will. I'll check.

One thing that not a lot of people are aware of with switch statements and expressions is that you can have multiple condition components to the switch without using a guard clause.

It looks like a tuple, basically:

```csharp
switch (value1, value2)
{
  case ({ BoolInValue1: true, StringInValue1: "1" }, 420 ):
  {
    // This will have matched if the `value1` object matched the pattern
    // And if value2 is a numeric type or implicitly convertible to a numeric type with the value 420.
  } 
}

You can have as many components as you need, and discards ( `_` ) can be used when you don't care about one or more of them in a given case.

Handy when you have control flow constructs that depend on both complex types and local variables, parameters, etc that wouldn't work without a guard clause otherwise.

And it's faster at run-time than using a guard clause, because it's compile-time constant, whereas guard clauses have to be evaluated _after_ the lookup/jump.

(Guard clauses still are still useful, though, if you _don't_ have a compile-time constant to match against.)